File: 백준/Ruby/13409. Black and White Boxes/Black and White Boxes.py
# 13409 Black and White Boxes
from sys import stdin


def input():
    return stdin.readline().strip()


def find_box_number(box_configuration: str) -> int:
    if not box_configuration:
        return 0
    box_list = [1 if i == "B" else -1 for i in box_configuration]
    sign = 1
    result = 0
    multiplier = 1 << 40
    # 처음이 W면 반대로 생각
    if box_list[0] == -1:
        box_list = [-i for i in box_list]
        sign = -1
    # 첫번째 색상 변경을 기준으로 생각
    first_color_change = -1
    for i, v in enumerate(box_list):
        if i == 0:
            continue
        if v != box_list[i - 1]:
            first_color_change = i - 1
            break
    # 색상 변경이 없을 경우 길이가 값이 됨
    if first_color_change == -1:
        result = multiplier * len(box_configuration) * sign
        return result
    # 색상 변경이 있을 경우 색상 변경 이후로는 0.5배씩 등비수열처럼 곱해짐
    result = multiplier * first_color_change
    box_list = box_list[first_color_change:]
    result += sum((multiplier >> i) * v for i, v in enumerate(box_list))
    return result * sign


def register_combination(sums_dict, status, size, numbers, offset):
    total_value, total_box_count = 0, 0
    for i in range(size):
        if not (status >> i) & 1:
            continue
        total_value += numbers[offset + i][0]
        total_box_count += numbers[offset + i][1]
    if total_value not in sums_dict or total_box_count > sums_dict[total_value]:
        sums_dict[total_value] = total_box_count


# 합이 0이 되는 조합 중 상자의 갯수를 가장 많이 가져갈 수 있는 조합 찾기
def find_max_box_number(numbers) -> int:
    n = len(numbers)
    n1, n2 = n // 2, n - n // 2
    sums_left = {}
    sums_right = {}
    for status in range(1, 1 << n1):
        register_combination(sums_left, status, n1, numbers, 0)
    for status in range(1, 1 << n2):
        register_combination(sums_right, status, n2, numbers, n1)
    max_box_count = 0
    for value, box_count in sums_left.items():
        if -value in sums_right:
            max_box_count = max(max_box_count, box_count + sums_right[-value])
    return max_box_count


def main():
    N = int(input())
    box_numbers = []
    for _ in range(N):
        box_configuration = input()
        box_numbers.append(
            (
                find_box_number(box_configuration),
                len(box_configuration),
            )
        )
    box_numbers.sort(key=lambda x: (x[0], -x[1]))
    result = find_max_box_number(box_numbers)
    print(result)


main()

File: 백준/Ruby/13409. Black and White Boxes/Black and White Boxes.kt
import java.util.*
import kotlin.collections.ArrayList
import kotlin.math.max

var noColorChange = -1
fun findBoxNumber(boxConfiguration: String): Long {
    if (boxConfiguration.isEmpty()) {
        return 0L
    }
    var boxList: ArrayList<Long> = ArrayList(List<Long>(0) { 0 })
    for (i in boxConfiguration.indices) {
        val isBlack: Long = if (boxConfiguration[i] == 'B') {
            1L
        } else {
            -1L
        }
        boxList.add(isBlack)
    }
    var sign = 1L
    var result: Long
    val multiplier = 1L shl 40
    if (boxList[0] == -1L) {
        sign = -1L
        for (i in boxList.indices) {
            boxList[i] = boxList[i] * -1L
        }
    }
    var firstColorChange = noColorChange
    for (i in boxList.indices) {
        if (i == 0) {
            continue
        }
        if (boxList[i] != boxList[i - 1]) {
            firstColorChange = i - 1
            break
        }
    }
    if (firstColorChange == noColorChange) {
        result = multiplier * boxConfiguration.length * sign
        return result
    }
    result = multiplier * firstColorChange
    for (i in boxList.indices) {
        if (i < firstColorChange) {
            continue
        }
        result += boxList[i] * (multiplier shr (i - firstColorChange))
    }
    return sign * result
}

fun registerCombination(
    sumsMap: MutableMap<Long, Long>,
    status: Long,
    size: Int,
    numbers: ArrayList<ArrayList<Long>>,
    offset: Int
) {
    var totalValue = 0L
    var totalBoxCount = 0L
    for (i in 0 until size) {
        if (((status shr i) and 1) == 0L) {
            continue
        }
        totalValue += numbers[offset + i][0]
        totalBoxCount += numbers[offset + i][1]
    }
    if (!sumsMap.containsKey(totalValue)) {
        sumsMap[totalValue] = totalBoxCount
    } else {
        sumsMap[totalValue] = max(sumsMap[totalValue]!!, totalBoxCount)
    }
}

fun findMaxBoxNumber(numbers: ArrayList<ArrayList<Long>>): Int {
    val n = numbers.size
    val n1 = n / 2
    val n2 = n - n / 2
    val sumsLeft: MutableMap<Long, Long> = mutableMapOf()
    val sumsRight: MutableMap<Long, Long> = mutableMapOf()
    for (status in 1L until (1L shl n1)) {
        registerCombination(sumsLeft, status, n1, numbers, 0)
    }
    for (status in 1L until (1L shl n2)) {
        registerCombination(sumsRight, status, n2, numbers, n1)
    }
    var maxBoxCount = 0
    sumsLeft.forEach {
        val (value, boxCount) = it
        if (sumsRight.containsKey(-value)) {
            maxBoxCount = max(maxBoxCount, (boxCount + sumsRight[-value]!!).toInt())
        }
    }
    return maxBoxCount
}

fun main() = with(Scanner(System.`in`)) {
    val n = nextLine().toInt()
    val boxNumbers = ArrayList<ArrayList<Long>>(List(n) { ArrayList(List(2) { 0L }) })
    for (i in 0 until n) {
        val boxConfiguration = nextLine()
        boxNumbers[i] = arrayListOf(findBoxNumber(boxConfiguration), boxConfiguration.length.toLong())
    }
    boxNumbers.sortBy { (it[0] shl 10) - it[1] }
    val result = findMaxBoxNumber(boxNumbers)
    println(result)
}